src/crimson/common/fixed_kv_node_layout.h

```diff
@@ -126,113 +138,291 @@ class FixedKVNodeLayout {
+  }
+
+
   FixedKVNodeLayout(char *buf) :
     buf(buf) {}
 
-  fixed_node_iter_t begin() {
-    return fixed_node_iter_t(
+  const_iterator begin() const {
+    return const_iterator(
+      this,
+      0);
+  }
+
+  const_iterator end() const {
+    return const_iterator(
+      this,
+      get_size());
+  }
+
+  iterator begin() {
+    return iterator(
       this,
       0);
   }
 
-  fixed_node_iter_t end() {
-    return fixed_node_iter_t(
+  iterator end() {
+    return iterator(
       this,
       get_size());
   }
 
-  fixed_node_iter_t iter_idx(uint16_t off) {
-    return fixed_node_iter_t(
+  const_iterator iter_idx(uint16_t off) const {
+    return const_iterator(
       this,
       off);
   }
 
-  fixed_node_iter_t find(K l) {
+  const_iterator find(K l) const {
     auto ret = begin();
     for (; ret != end(); ++ret) {
       if (ret->get_key() == l)
 	break;
     }
     return ret;
   }
+  iterator find(K l) {
+    const auto &tref = *this;
+    return iterator(this, tref.find(l).offset);
+  }
 
-  fixed_node_iter_t get_split_pivot() {
-    return iter_idx(get_size() / 2);
+  const_iterator lower_bound(K l) const {
+    auto ret = begin();
+    for (; ret != end(); ++ret) {
+      if (ret->get_key() > l)
+	break;
+    }
+    return ret;
+  }
+  iterator lower_bound(K l) {
+    const auto &tref = *this;
+    return iterator(this, tref.lower_bound(l).offset);
   }
 
-private:
-  KINT *get_key_ptr() {
-    return layout.template Pointer<1>(buf);
+  const_iterator upper_bound(K l) const {
+    auto ret = begin();
+    for (; ret != end(); ++ret) {
+      if (ret->get_key() > l)
+	break;
+    }
+    return ret;
+  }
+  iterator upper_bound(K l) {
+    const auto &tref = *this;
+    return iterator(this, tref.upper_bound(l).offset);
   }
 
-  VINT *get_val_ptr() {
-    return layout.template Pointer<2>(buf);
+  const_iterator get_split_pivot() const {
+    return iter_idx(get_size() / 2);
   }
 
-public:
   uint16_t get_size() const {
     return *layout.template Pointer<0>(buf);
   }
 
-  void set_size(uint16_t size) {
-    *layout.template Pointer<0>(buf) = size;
-  }
-
   constexpr static size_t get_capacity() {
     return CAPACITY;
   }
 
-  /**
-   * copy_from_foreign
-   *
-   * Copies entries from [from_src, to_src) to tgt.
-   *
-   * tgt and from_src must be from different nodes.
-   * from_src and to_src must be from the same node.
-   */
-  static void copy_from_foreign(
-    fixed_node_iter_t tgt,
-    fixed_node_iter_t from_src,
-    fixed_node_iter_t to_src) {
-    assert(tgt->node != from_src->node);
-    assert(to_src->node == from_src->node);
-    memcpy(
-      tgt->get_val_ptr(), from_src->get_val_ptr(),
-      to_src->get_val_ptr() - from_src->get_val_ptr());
-    memcpy(
-      tgt->get_key_ptr(), from_src->get_key_ptr(),
-      to_src->get_key_ptr() - from_src->get_key_ptr());
-  }
+  bool operator==(const FixedKVNodeLayout &rhs) const {
+    if (get_size() != rhs.get_size()) {
+      assert(false);
```

Comment:
do you mean, it's not expected that `lhs != rhs` ?

Reply:
Mistake, left over from debugging, will remove, thanks.